plugins/bigquery/dbt/include/bigquery/macros/materializations/incremental.sql

```diff
@@ -1,160 +1,169 @@
+    {%- set unique_key = config.get('unique_key') -%}
+    {%- set full_refresh_mode = (should_full_refresh()) -%}
+
+    {%- set target_relation = this %}
+    {%- set existing_relation = load_relation(this) %}
+    {%- set tmp_relation = make_temp_relation(this) %}
+
+    {#-- Validate early so we don't run SQL if the strategy is invalid --#}
+    {% set strategy = dbt_bigquery_validate_get_incremental_strategy(config) -%}
+
+    {%- set raw_partition_by = config.get('partition_by', none) -%}
+    {%- set partition_by = adapter.parse_partition_by(raw_partition_by) -%}
+    {%- set partitions = config.get('partitions', none) -%}
+    {%- set is_partition_filter_required = config.get('require_partition_filter', false) -%}
+    {%- set cluster_by = config.get('cluster_by', none) -%}
+
+    {{ run_hooks(pre_hooks) }}
+
+    {% if existing_relation is none %}
+        {% set build_sql = create_table_as(False, target_relation, sql) %}
+    {% elif existing_relation.is_view %}
+        {#-- There's no way to atomically replace a view with a table on BQ --#}
+        {{ adapter.drop_relation(existing_relation) }}
+        {% set build_sql = create_table_as(False, target_relation, sql) %}
+    {% elif full_refresh_mode %}
+        {#-- If the partition/cluster config has changed, then we must drop and recreate --#}
+        {% if not adapter.is_replaceable(existing_relation, partition_by, cluster_by) %}
+            {% do log("Hard refreshing " ~ existing_relation ~ " because it is not replaceable") %}
+            {{ adapter.drop_relation(existing_relation) }}
         {% endif %}
+        {% set build_sql = create_table_as(False, target_relation, sql) %}
+    {% else %}
+        {% set dest_columns = adapter.get_columns_in_relation(existing_relation) %}
+
+        {#-- if partitioned, use BQ scripting to get the range of partition values to be updated --#}
+        {% if strategy == 'insert_overwrite' %}
+
+            {% set missing_partition_msg -%}
+                The 'insert_overwrite' strategy requires the `partition_by` config.
+            {%- endset %}
+            {% if partition_by is none %}
+                {% do exceptions.raise_compiler_error(missing_partition_msg) %}
+            {% endif %}
+
+            {% set build_sql = bq_insert_overwrite(
+                tmp_relation,
+                target_relation,
+                sql,
+                unique_key,
+                partition_by,
+                partitions,
+                dest_columns) %}
+
+        {% else %}
+            {#-- wrap sql in parens to make it a subquery --#}
+            {%- set source_sql -%}
+                (
+                    {{ sql }}
+                )
+            {%- endset -%}
+
+            {% set predicates = [] %}
+            {% if is_partition_filter_required %}
+                {%- set partition_filter -%}
+                    (DBT_INTERNAL_DEST.{{ partition_by.field }} is not null or DBT_INTERNAL_DEST.{{ partition_by.field }} is null)
```

Comment:
I wasn't sure if there was a better way to get the `DBT_INTERNAL_DEST` alias in there than just typing it out. I tried `{{ partition_by.render(alias='DBT_INTERNAL_DEST') }}` but that renders to `timestamp_trunc(my_partition_col, date)` when partitioning by date on a timestamp col, and BigQuery doesn't like that as a partition filter.

Reply:
> and BigQuery doesn't like that as a partition filter

Really? It works as a `partition by` expression, but not as a partition-pruning filter? That's... too bad.

The approach you've taken here seems fine by me